packages/graphql/src/types.ts

```diff
@@ -42,6 +47,7 @@ export interface Context {
     jwt?: JwtPayload;
     auth?: AuthContext;
     queryOptions?: CypherQueryOptions;
+    executionContext?: ExecutionContext;
```

Comment:
The typing decisions which have been made here are a bit awkward - only one of `Driver`, `Session` and `Transaction` will be used, so perhaps this should be a union? `executionContext` also clashes with the `driver` field somewhat, so doing this would allow us to potentially depreciate the `driver` field in the next major release.

```suggestion
    executionContext?: Driver | Session | Transaction;
```

Reply:
Thanks for your reply! I agree that the typing is a bit awkward, and the API is a bit cryptic. 

The execute function uses more than one of them though (Driver, Session and Transaction). 

The driver is used [here](https://github.com/neo4j/graphql/blob/d4cba9a4e2bec99f4bb7b07e6dfd99acd2cace8e/packages/graphql/src/utils/execute.ts#L82-L88), but I am not sure if that part is actually important, or if it could be skipped in cases where a driver is not defined?

As it stands the Session is used to return a bookmark [here](https://github.com/neo4j/graphql/blob/d4cba9a4e2bec99f4bb7b07e6dfd99acd2cace8e/packages/graphql/src/utils/execute.ts#L124). With the new changes it is possible to specify an executionContext, so that the transaction will not be commited automatically. This means that the returned bookmark will not include the open transaction. I guess that can be a bit counterintuitive.

I can think of a couple of options for solving this:

-  We replace executionContext with something named just `transaction`, and then let the returned bookmark be null in cases where we only have a transaction and no session.
-  Or we can do it like you say. The API only lets you provide one of Driver, Session and Transaction, and then we just skip the parts of the code that requires a Session or Driver in the cases where they are not defined. This solution would also mean that the returned bookmark can be null.

I like the second option better, but it would also require changes to larger parts of the code which I am not that familiar with.

Reply:
> The driver is used [here](https://github.com/neo4j/graphql/blob/d4cba9a4e2bec99f4bb7b07e6dfd99acd2cace8e/packages/graphql/src/utils/execute.ts#L82-L88), but I am not sure if that part is actually important, or if it could be skipped in cases where a driver is not defined?

No need to worry too much about that, the functionality is pretty flaky at the best of times anyway. We actually want to swap this out for putting the same information in the transaction metadata - feel free to add this in if you feel up to it!!

> As it stands the Session is used to return a bookmark [here](https://github.com/neo4j/graphql/blob/d4cba9a4e2bec99f4bb7b07e6dfd99acd2cace8e/packages/graphql/src/utils/execute.ts#L124). With the new changes it is possible to specify an executionContext, so that the transaction will not be commited automatically. This means that the returned bookmark will not include the open transaction. I guess that can be a bit counterintuitive.

We only really need to worry about returning the bookmark if we committed the transaction - otherwise it is the users responsibility to deal with the effects of causal clustering.

> * Or we can do it like you say. The API only lets you provide one of Driver, Session and Transaction, and then we just skip the parts of the code that requires a Session or Driver in the cases where they are not defined. This solution would also mean that the returned bookmark can be null.

I like this option, and I can think of very few downsides!! No issues making a currently non-nullable field nullable.

Reply:
Ok, I have reimplemented it the way we discussed :+1: